### ISSUE TITLE
[gatsby-link] TypeError useRef and Link component

### ISSUE THREAD

**Issue by ollelauribostrom** [gatsbyjs/gatsby#12014]
## Description
I'm trying to use the new useRef hook to create a reference to a Link component. This results in an error saying `TypeError: this.props.innerRef is not a function`.

### Steps to reproduce
```js
  const linkRef = useRef(null);
  return (
    <Link
      ref={linkRef}
      to="https://github.com/gatsbyjs/gatsby-starter-hello-world"
    >
      Hello world!
    </Link>
  );
```
Full example project [here](https://github.com/ollelauribostrom/gatsby-link-useref). Same exact result if passing the `innerRef` props instead.

### Expected result
I expect the mutable ref object to reference the rendered `<a>`-tag

### Actual result
Error thrown from [here](https://github.com/gatsbyjs/gatsby/blob/fd12550dae58becdc52057d8b43f1eb6960a6883/packages/gatsby-link/src/index.js#L71).
```
Uncaught TypeError: this.props.innerRef is not a function
    at GatsbyLink.handleRef (index.js:117)
    at GatsbyLink.handleRef (react-hot-loader.development.js:652)
    at commitAttachRef (react-dom.development.js:17291)
    at commitAllLifeCycles (react-dom.development.js:18537)
    at HTMLUnknownElement.callCallback (react-dom.development.js:149)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:199)
    at invokeGuardedCallback (react-dom.development.js:256)
    at commitRoot (react-dom.development.js:18744)
    at react-dom.development.js:20214
    at Object.unstable_runWithPriority (scheduler.development.js:255)
    at completeRoot (react-dom.development.js:20213)
    at performWorkOnRoot (react-dom.development.js:20142)
    at performWork (react-dom.development.js:20050)
    at performSyncWork (react-dom.development.js:20024)
    at requestWork (react-dom.development.js:19893)
    at scheduleWork (react-dom.development.js:19707)
    at scheduleRootUpdate (react-dom.development.js:20368)
    at updateContainerAtExpirationTime (react-dom.development.js:20396)
    at updateContainer (react-dom.development.js:20453)
    at ReactRoot../node_modules/react-dom/cjs/react-dom.development.js.ReactRoot.render (react-dom.development.js:20749)
    at react-dom.development.js:20886
    at unbatchedUpdates (react-dom.development.js:20255)
    at legacyRenderSubtreeIntoContainer (react-dom.development.js:20882)
    at render (react-dom.development.js:20951)
    at app.js:56
```

### Environment
  System:
    OS: OS X El Capitan 10.11.1
    CPU: (4) x64 Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
    Shell: 5.0.8 - /bin/zsh
  Binaries:
    Node: 11.10.0 - /usr/local/bin/node
    Yarn: 1.12.3 - /usr/local/bin/yarn
    npm: 6.7.0 - /usr/local/bin/npm
  Languages:
    Python: 2.7.10 - /usr/bin/python
  Browsers:
    Chrome: 72.0.3626.119
    Firefox: 63.0.3
    Safari: 9.0.1
  npmPackages:
    gatsby: ^2.1.17 => 2.1.17 
  npmGlobalPackages:
    gatsby-cli: 2.4.8

**Comment by yogeshkotadiya** [comment]
`ref` expects a function in the `Link` component and `gatsby-link` is just a wrapper around `reach-router's` Link component.
If you want to know more about it check the API for reach/router.
[https://reach.tech/router/api/Link](https://reach.tech/router/api/Link)

**Comment by ollelauribostrom** [comment]
The `reach-router's` [Link component](https://github.com/reach/router/blob/b60e6dd781d5d3a4bdaaf4de665649c0f6a7e78d/src/index.js#L374) does actually allow you to pass a `MutableRefObject` instead of a function. That seems like to correct approach when using the new hooks API (see the docs [here](https://reactjs.org/docs/hooks-reference.html#useref)).  But I'm guessing that `gatsby-link` requires a function to be able to both setup observer functionality and pass back the reference? 🤔

**Comment by freiksenet** [comment]
We haven't updated gatsby-link to use React>=16.4 API yet, so you can only use function refs through innerRef function. `ref` isn't even an allowed prop type in Gatsby link component. We definitely do want to support it though, do you want to make a PR to do that?

**Comment by ollelauribostrom** [comment]
@freiksenet I'm definitely up for that. Does that mean we should allow both `ref` and `innerRef`?

**Comment by freiksenet** [comment]
@ollelauribostrom yes, let's match reach-router API.

**Comment by ollelauribostrom** [comment]
@freiksenet Had a go at this in #12174 🙂

**Comment by yogeshkotadiya** [comment]
Merged in #12174
